### PR TITLE
Processes as entities

### LINKED ISSUE #1: Design Proposal: Processes as Entities
I was planning to develop a similar crate as this one and therefore thought a lot about potential API designs.

One design that I think is worth exploring is to treat processes as entities in Bevy.
So to spawn a process, you would spawn an entity with a particular component (e.g. `Process` or `Command` component) and to kill a process you would despawn this entity.
The plugin's systems would take care of creating the actual processes based on the `Created` events and kill the process (or if the process ends it would despawn the entity).

I think this design provides a couple of advantages:

- It uses Bevy-native concepts like spawning and despawning entities.
- If you have many different processes, you can categorize and filter them with marker components, etc. You can use the query mechanisms to organize your processes.
- You can attach additional data to the processes simply by adding additional components to the entities.
- Looking up specific processes becomes faster, as lookup by entity ID is efficient in Bevy. The PID can still be provided in the form of a component attached to the entity.

Of course there may also be some disadvantages that I haven't thought of yet. For example, despawning the entity once the process has finished might be harder than I think.
Still, I think it might be worth it to explore this design.

### PR DESCRIPTION
**Pull request by TimJentzsch**
Closes #1.

This PR explores modeling processes as entities.

We use two components:
- `LocalCommand`, which contains the `std::process::Command` to run. You spawn this component to run the command.
- `Process`, which contains the running process, the output buffer, and the task which reads the output.

To spawn a new process, you don't use events anymore, but Bevy's `Commands`:

```rs
fn startup(mut commands: Commands) {
    let mut cmd = std::process::Command::new("sh");
    cmd.args(["-c", "echo Sleeping for 1s && sleep 1 && echo Done"]);

    let id = commands.spawn(LocalCommand::new(cmd)).id();
    println!("Spawned the command as entity {id:?}");
}
```

I debated also adding an API to use an event to spawn the process, but that's actually equally or even more verbose than just using `Commands`.

The events now only store the `Entity` of the corresponding `Process`.
To obtain the PID, Command that was run, and other information, you can simply query for `Command` and/or `Process` and search for the entity.
This should be a lot more performant than constantly cloning strings and such.

We also don't need the `ActiveProcessMap` anymore, we just use the ECS as a collection of our processes.

Using the ECS directly also makes it easier to manage larger collection of processes.
You can simply add marker components or other metadata to the corresponding entities and then filter for that in your queries.

# Open Questions

- Getting other information (PID, command string representation) for events is now a bit more annoying, as you have to run a query as well. On the other hand, it should scale a lot better and be more performant. Is this an acceptable trade-off? Of course, we could still add more info to the events, at the cost of performance.
- In theory you could despawn the entity without killing the process first. I wonder if that could "leak" the processes and leave them running, or if dropping the Process will automatically kill them...

### CODE REVIEW

**Comment by TimJentzsch** [comment]
I think that this model also provides a nice future path for this crate, as we can develop "add-ons" to further modify the behavior of command execution.

For example, we can have a `LocalCommandChain` component which executes multiple commands after one another.
Then we can add a `Retry` component which controls what happens when the command fails and can simply retry a couple of times or use an exponential backoff strategy, etc.
Or maybe a `Repeater` component which executes the command in an infinite loop...

I think having a modular setup like this is a great path forward as we can allow the user to customize how they want their commands to be executed without having a giant monolithic setup

**Comment by edouardpoitras** [comment]
Nice! I really like this. I've been sick, but will review and test over the
weekend. Thanks 🙏

File: src/lib.rs [hunk 1]
```diff
@@ -105,101 +102,93 @@ impl Plugin for BevyLocalCommandsPlugin {
...
     for kill_shell_command in kill_process_event.read() {
-        let pid = kill_shell_command.0;
+        let entity: Entity = kill_shell_command.0;
 
-        if let Some(active_process) = active_process_map.0.get_mut(&pid) {
+        if let Ok(mut process) = query.get_mut(entity) {
             // Kill the process
             // TODO: Handle unwrap
-            active_process.process.kill().unwrap();
+            process.process.kill().unwrap();
```
Comment (edouardpoitras) [inline src/lib.rs]: Might as well use your Process.kill() implementation here and shorten it to `process.kill().unwrap()`

File: src/lib.rs [hunk 2]
```diff
@@ -105,101 +102,93 @@ impl Plugin for BevyLocalCommandsPlugin {
...
-        process_started_event.send(ProcessStarted { command, pid });
-
-        active_process_map.0.insert(pid, active_process);
+    for (entity, mut local_command) in query.iter_mut() {
+        match spawn_process(&mut local_command.command) {
+            Ok(process) => {
+                commands.entity(entity).insert(process);
+            },
+            Err(_) => {
+                process_error_event.send(ProcessError {
```
Comment (edouardpoitras) [inline src/lib.rs:111]: If we're despawning entities when the process completes, do we want to do the same when the process fails?

File: src/lib.rs (hunk 1 above)
Comment (edouardpoitras) [reply src/lib.rs]: Same question regarding despawning here - do we want to despawn the entity?

EDIT: Ah never mind, I think the entity will be despawned once the process is completed in the handle_completed_process function.

**Comment by edouardpoitras** [comment]
Looks pretty solid to me. I much prefer this new approach. Looking forward to adding some of those "add-ons" you mentioned.

Does it make sense to just remove the KillProcess event now and have the user perform a query to fetch the process and kill it themselves? If they need the entity ID anyways, the user will probably end up querying it regardless?

If you get a chance, updating the examples in the README would save me the trouble.
Also, you can go ahead and bump up the version to 0.2.5.

Thanks!

File: src/lib.rs (hunk 1 above)
Comment (TimJentzsch) [reply src/lib.rs]: Correct, the entity is despawned in the completion handler

File: src/lib.rs (hunk 2 above)
Comment (TimJentzsch) [reply src/lib.rs:111]: Hmm, good question.

The automatic despawning is one area that I'm unsure about, as the user could also have attached additional information to the entity in different components.
But maybe that's something we can add as a configuration option later.
Or we add a marker component which prevents automatic despawning, similar to the other add on ideas.
Then the user has fine-grained control if they need it.

**Comment by TimJentzsch** [comment]
> Does it make sense to just remove the KillProcess event now and have the user perform a query to fetch the process and kill it themselves? If they need the entity ID anyways, the user will probably end up querying it regardless?

Good point, I'll remove the kill event

**Comment by TimJentzsch** [comment]
> Also, you can go ahead and bump up the version to 0.2.5.

I believe we'll have to bump to `0.3.0` as this is a breaking change.
Otherwise, `cargo` might automatically update the dependency of projects and it will break their build.
See <https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field> for reference

**Comment by edouardpoitras** [comment]
Looks good to me!

**Comment by edouardpoitras** [comment]
FYI - regarding your 2nd open question, I did some testing and this is what happens on my Linux box if you despawn the entity before killing the process:
- Process continues in the background
- No more ProcessOutputEvent
- No more ProcessCompletedEvent
- Process is killed if Bevy is stopped

**Comment by TimJentzsch** [comment]
> FYI - regarding your 2nd open question, I did some testing and this is what happens on my Linux box if you despawn the entity before killing the process:
> 
>     * Process continues in the background
> 
>     * No more ProcessOutputEvent
> 
>     * No more ProcessCompletedEvent
> 
>     * Process is killed if Bevy is stopped

Thanks for investigating this!
This seems like a potential footgun, I wonder if we can prevent this somehow.
Otherwise, we should probably document this behavior.

Maybe something for a future PR though.

**Pull request merged**